MdeModulePkg/Bus/Pci/PciBusDxe/PciEnumeratorSupport.c

```diff
@@ -207,6 +207,155 @@ PciPciDeviceInfoCollector (
+    }
+
+    *MaxPayloadSize = PciIoDevice->MaxPayloadSize;
+  } else {
+    *MaxPayloadSize = MAX_UINT8;
+  }
+
+  //
+  // Recurse into each child to find the max payload size supported.
+  //
+  CurrentLink = PciIoDevice->ChildList.ForwardLink;
+  while (CurrentLink != NULL && CurrentLink != &PciIoDevice->ChildList) {
+    Child  = PCI_IO_DEVICE_FROM_LINK (CurrentLink);
+    Status = PciGetMaxPayloadSize (Child, &ChildMps);
+    if (EFI_ERROR (Status)) {
+      return Status;
+    }
+
+    if (ChildMps < *MaxPayloadSize) {
+      *MaxPayloadSize = ChildMps;
+    }
+
+    CurrentLink = CurrentLink->ForwardLink;
+  }
+
+  if (*MaxPayloadSize == MAX_UINT8) {
+    return EFI_UNSUPPORTED;
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Sets the PCIe Max PayloadSize for the provided device and it's children.
+
+  @param[in]  PciIoDevice       The PCI IO Device to set the MPS for.
+  @param[out] MaxPayloadSize    The MPS to set.
+
+  @retval   EFI_SUCCESS         The MPS ws set for the device and it's children.
+  @retval   EFI_UNSUPPORTED     MPS not supported by provided device or its children.
+  @retval   Other               MPS could not be read or written.
+
+**/
+EFI_STATUS
+PciProgramMps (
+  IN PCI_IO_DEVICE  *PciIoDevice,
+  IN UINT8          MaxPayloadSize
+  )
+{
+  LIST_ENTRY                   *CurrentLink;
+  PCI_IO_DEVICE                *Child;
+  EFI_STATUS                   Status;
+  PCI_REG_PCIE_DEVICE_CONTROL  DeviceControl;
+
+  //
+  // Skip the root bridge.
+  //
+
+  if (PciIoDevice->Parent != NULL) {
+    if (!PciIoDevice->IsPciExp) {
+      return EFI_UNSUPPORTED;
+    }
+
+    Status = PciIoDevice->PciIo.Pci.Read (
+                                      &PciIoDevice->PciIo,
+                                      EfiPciIoWidthUint16,
+                                      PciIoDevice->PciExpressCapabilityOffset + EFI_PCIE_CAPABILITY_DEVICE_CONTROL,
+                                      1,
+                                      &DeviceControl.Uint16
+                                      );
+    if (EFI_ERROR (Status)) {
+      return Status;
+    }
+
+    if (DeviceControl.Bits.MaxPayloadSize != MaxPayloadSize) {
+      DeviceControl.Bits.MaxPayloadSize = MaxPayloadSize;
+      DEBUG ((
+        DEBUG_INFO,
+        "%a: %02x %02x %02x Setting MPS: %x\n",
+        __FUNCTION__,
+        PciIoDevice->BusNumber,
+        PciIoDevice->DeviceNumber,
+        PciIoDevice->FunctionNumber,
+        MaxPayloadSize
+        ));
+
+      Status = PciIoDevice->PciIo.Pci.Write (
+                                        &PciIoDevice->PciIo,
+                                        EfiPciIoWidthUint16,
+                                        PciIoDevice->PciExpressCapabilityOffset + EFI_PCIE_CAPABILITY_DEVICE_CONTROL,
+                                        1,
+                                        &DeviceControl.Uint16
+                                        );
+      if (EFI_ERROR (Status)) {
+        return Status;

```

Comment:
Let's say that a root bridge has another bridge behind it, Ann's beyond that bridge, there are two devices. One device fails to program max payload (I don't see how that would happen because most pci devices just discard writes to read only registers, but for the sake of this scenario, let's say it does). We would leave one additional device that would not have the correct payload size.  Maybe we should output a debug message and continue going through the child devices?

An an ascii example the confusing paragraph above:
Rp0 - Bridge0 - device 1
              - device 2 <- write error
              - device 3 <- skipped by return

Reply:
Yeah, I considered the failure scenarios for a bit, but I'm not sure the full implications of this in practice. I'm going to talk to the OS folks a bit on this. The alternative would be to re-set all values down to the baseline 0 value.

Reply:
Looking into it more, this should only happen if the device's information is bad or we are in the critical failure path anyways. I will log an error and continue on  the other devices.